### PR TITLE
Adding elastic loggging for drift detector data

### PR DESCRIPTION
**Pull request by SachinVarghese**
**What this PR does / why we need it**:
This PR adds elasticsearch logging for alibi specific drift detector metrics in the request logger component

**Special notes for your reviewer**:
There are no breaking changes introduced but does involve creating a new index pattern, so the elastic token needs the required permissions
**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
Adds new log entry with drift detector data

chrome-extension___aejoelaoggembcahagimdiliamlcdmfm_index html (1)

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **sachinvarghese**
You can assign the PR to them by writing `/assign @sachinvarghese` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by michaelcheah — APPROVED** [review]
lgtm

Couple minor nits but considering this code is going to be fully refactored, not particularly fussed about them

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *[michaelcheah](https://github.com/SeldonIO/seldon-core/pull/3560#pullrequestreview-744903912 "Approved")*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **sachinvarghese**
You can assign the PR to them by writing `/assign @sachinvarghese` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/seldon-request-logger/app/default_logger.py [hunk 1]
```diff
@@ -206,6 +206,42 @@ def process_and_update_elastic_doc(
...
+            ):
+                content_dist = np.array(item_body[message_type]["data"]["p_val"])
+                x = np.expand_dims(content_dist, axis=0)
+                item_body[message_type]["data"]["drift_type"] = "feature"
+                item_body[message_type]["data"]["p_val"] = createElelmentsArray(x, None, namespace, serving_engine, inferenceservice_name, endpoint_name, "request")
+        detectorName=None
+        if item_body[message_type]["ce-source"].startswith("io.seldon.serving."):
+            detectorName =  item_body[message_type]["ce-source"][len("io.seldon.serving."):]
+        elif item_body[message_type]["ce-source"].startswith("org.kubeflow.serving."):
+            detectorName =  item_body[message_type]["ce-source"][len("org.kubeflow.serving."):]
```
[the remark this answers is not in the document]
Comment (SachinVarghese) [reply components/seldon-request-logger/app/default_logger.py]: Updated now

File: components/seldon-request-logger/app/log_helper.py [hunk 2]
```diff
@@ -34,42 +34,46 @@ def extract_request_id(headers):
         request_id = headers.get(CLOUD_EVENT_ID)
     return request_id
 
-
-def build_index_name(headers):
+def build_index_name(headers, prefix = None, suffix = True, nameOverride = None):
```
[the remark this answers is not in the document]
Comment (SachinVarghese) [reply components/seldon-request-logger/app/log_helper.py]: Updated now

**Review by axsaucedo — CHANGES_REQUESTED** [review]
Looks godo, added a couple of comments - woudl be good to cover any key thigns on the UPGRADING.md page

File: components/seldon-request-logger/app/log_helper.py [hunk 3]
```diff
@@ -34,42 +34,46 @@ def extract_request_id(headers):
...
     else:
         index_name = index_name + "-" + namespace
-    inference_service_name = clean_header(INFERENCESERVICE_HEADER_NAME, headers)
-    # won't get inference service name for older kfserving versions i.e. prior to https://github.com/kubeflow/kfserving/pull/699/
-    if not inference_service_name:
-        inference_service_name = clean_header(MODELID_HEADER_NAME, headers)
-
+    if name_override is None:
+        inference_service_name = clean_header(INFERENCESERVICE_HEADER_NAME, headers)
+        # won't get inference service name for older kfserving versions i.e. prior to https://github.com/kubeflow/kfserving/pull/699/
```
[the remark this answers is not in the document]
Comment (SachinVarghese) [reply components/seldon-request-logger/app/log_helper.py:61]: Will create a separate issue for docs

File: components/seldon-request-logger/app/log_mapping.py [hunk 4]
```diff
@@ -206,7 +206,7 @@ def fetch_metadata(namespace, serving_engine, inferenceservice_name, predictor_n
             return None
 
         print('prediction schema for '+namespace+'/'+inferenceservice_name)
-        print(model_metadata.models[0].prediction_schema)
+        # print(model_metadata.models[0].prediction_schema)
```
[the remark this answers is not in the document]
Comment (SachinVarghese) [reply components/seldon-request-logger/app/log_mapping.py]: Fixed now

File: components/seldon-request-logger/app/log_helper.py [hunk 5]
```diff
@@ -34,42 +34,46 @@ def extract_request_id(headers):
...
-
-def build_index_name(headers):
+def build_index_name(headers, prefix = None, suffix = True, name_override = None):
     # use a fixed index name if user chooses to do so
     index_name = os.getenv("INDEX_NAME")
     if index_name:
         return index_name
 
+    if prefix is None:
+        prefix = "inference"
```
[the remark this answers is not in the document]
Comment (SachinVarghese) [reply components/seldon-request-logger/app/log_helper.py]: Good point, fixed now

File: components/seldon-request-logger/app/default_logger.py [hunk 6]
```diff
@@ -206,6 +206,43 @@ def process_and_update_elastic_doc(
                 elastic_object, message_type, item_body, item_request_id, index_name
             )
             index = index + 1
+    elif "data" in new_content_part and message_type == "drift":
```
[the remark this answers is not in the document]
Comment (SachinVarghese) [reply components/seldon-request-logger/app/default_logger.py:209]: This is kept same as outliers, if data not present it already throws error. So I feel this should be okay

**Comment by axsaucedo** [comment]
NIce
/approve

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/3560#issuecomment-911792417 "Approved")*, *[michaelcheah](https://github.com/SeldonIO/seldon-core/pull/3560#pullrequestreview-744903912 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**